File: hackerrank/day15/day15.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class Solution:
    def insert(self, head, data):
        new_node = Node(data)

        # Si la lista está vacía
        if head is None:
            return new_node

        # Recorrer hasta el último nodo
        current = head
        while current.next is not None:
            current = current.next

        # Insertar el nuevo nodo al final
        current.next = new_node

        return head

    def display(self, head):
        current = head

        while current is not None:
            print(current.data, end=" ")
            current = current.next

        print()


def main():
    mylist = Solution()
    head = None

    cantidad = int(input("¿Cuántos números desea insertar? "))

    for i in range(cantidad):
        dato = int(input(f"Número #{i + 1}: "))
        head = mylist.insert(head, dato)

    print("\nLista enlazada:")
    mylist.display(head)


if __name__ == "__main__":
    main()

File: hackerrank/day15-1/day15.js
const readline = require("readline");

class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

class Solution {
    insert(head, data) {
        const newNode = new Node(data);

        if (head === null) {
            return newNode;
        }

        let current = head;

        while (current.next !== null) {
            current = current.next;
        }

        current.next = newNode;

        return head;
    }

    display(head) {
        let current = head;

        while (current !== null) {
            process.stdout.write(current.data + " ");
            current = current.next;
        }

        console.log();
    }
}

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

const myList = new Solution();
let head = null;

rl.question("¿Cuántos números desea insertar? ", (cantidad) => {
    let T = parseInt(cantidad);
    let contador = 0;

    function leerNumero() {
        if (contador === T) {
            console.log("\nLista enlazada:");
            myList.display(head);
            rl.close();
            return;
        }

        rl.question(`Número #${contador + 1}: `, (valor) => {
            head = myList.insert(head, parseInt(valor));
            contador++;
            leerNumero();
        });
    }

    leerNumero();
});